### DISCUSSION TITLE
How to flash code using VSCode & platformio on Atmega32U4 custim board

### DISCUSSION THREAD

**Discussion by michbichler** [avrdudes/avrdude#1366] [General]
Hello Dears,

I woul like to flash a working code (tried it out on a leonardo board) on a slef made custum board with a Atmega32U4 Controller with the bootloader fresh from the factory). 
Unfortunately just this error message appears:

![Screenshot](https://github.com/avrdudes/avrdude/assets/114282671/64a0dec0-4988-4d99-8bbf-72b085c93c88)

Do you have a idea, whats wrong here?
best regards,
Michael

**Comment by MCUdude** [comment]
If it is a custom board, are you using the built-in DFU bootloader that's factory loaded on the ATmega32U4?

Can you paste the contents of your platformio.ini file?

**Reply by michbichler** [reply]
Yes, the DFU bootloader is on it, I guess. Using flip I have a connection with the chip and can, for example, erase the chip.
Here is the content of the platformio.ini file:

``` ini
[env:my_board]
platform = atmelavr
board = my_board
framework = arduino
lib_deps = 
	SPI
	adafruit/Adafruit BusIO@^1.14.1
	Wire
	adafruit/Adafruit GFX Library@^1.10.1
	nkawu/TFT 22 ILI9225@^1.4.5
	bblanchon/ArduinoJson@^6.20.0
```

**Reply by MCUdude** [reply]
Try this environment instead:

```ini
[env:my_board]
platform = atmelavr
board = my_board
framework = arduino
upload_protocol = custom
upload_flags =
  -C$PROJECT_PACKAGES_DIR/tool-avrdude/avrdude.conf
  -p$BOARD_MCU
  -cflip1
upload_command = avrdude $UPLOAD_FLAGS -U flash:w:$SOURCE:i

lib_deps = 
	SPI
	adafruit/Adafruit BusIO@^1.14.1
	Wire
	adafruit/Adafruit GFX Library@^1.10.1
	nkawu/TFT 22 ILI9225@^1.4.5
	bblanchon/ArduinoJson@^6.20.0
```

And either click the upload button or run `pio run -t upload -e my_board`

You'll of course have to make sure the board is actually in bootloader mode for it to be able to communicate with Avrdude.

**Reply by michbichler** [reply]
I think, it worked! Thank you.
![Screenshot](https://github.com/avrdudes/avrdude/assets/114282671/9368f2ce-41a4-4377-a3df-19e7f257a5aa)

**Reply by michbichler** [reply]
I just don't know, what that error means.

**Reply by MCUdude** [reply]
It's using the older Avrdude v6.3.

Try replacing upload_flags with this. If you're still getting warnings, try removing `-u`.

```ini
upload_flags =
  -C$PROJECT_PACKAGES_DIR/tool-avrdude/avrdude.conf
  -p$BOARD_MCU
  -cflip1
  -u
  -s
```

**Reply by michbichler** [reply]
Unfortunately there is the same error using the new upload flags, even when I remove -u.

**Reply by MCUdude** [reply]
There's nothing to worry about really. Avrdude 6.3 is only telling you that it's trying to read the fuses (safemode is enabled by default), but a bootloader can't do this.

However, there should be an easy solution to this, disable safemode, which was removed in Avrdude 7.0 anyways.

Can you paste the entire output when you run:

`pio run -t upload -e my_board -v` ?

I need to see the verbose PlatformIO output to know why safemode is still enabled

**Reply by michbichler** [reply]
Processing my_board (platform: atmelavr; board: my_board; framework: arduino; upload_protocol: custom; upload_command: avrdude $UPLOAD_FLAGS -U flash:w:$SOURCE:i; upload_flags: -C$PROJECT_PACKAGES_DIR/tool-avrdude/avrdude.conf, -p$BOARD_MCU, -cflip1, -s; lib_deps: SPI, adafruit/Adafruit BusIO@^1.14.1, Wire, adafruit/Adafruit GFX Library@^1.10.1, nkawu/TFT 22 ILI9225@^1.4.5, bblanchon/ArduinoJson@^6.20.0)
----------------------------------------------------------------------------------------------------------------------------------------------CONFIGURATION: https://docs.platformio.org/page/boards/atmelavr/my_board.html
PLATFORM: Atmel AVR (4.1.0) > SparkFun Pro Micro 5V/16MHz
HARDWARE: ATMEGA32U4 16MHz, 2.50KB RAM, 28KB Flash       
DEBUG: Current (simavr) External (simavr)
PACKAGES:
 - framework-arduino-avr @ 5.1.0
 - tool-avrdude @ 1.60300.200527 (6.3.0)
 - toolchain-atmelavr @ 1.70300.191015 (7.3.0)
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 9 compatible libraries
Scanning dependencies...
Dependency Graph
|-- SPI @ 1.0 (License: Unknown, Path: C:\Users\micha\.platformio\packages\framework-arduino-avr\libraries\SPI)
|-- Adafruit BusIO @ 1.14.1 (License: Unknown, Path: C:\Users\micha\OneDrive\Berglerwerk\Berglerwerk_Github\Pen2.0\Pen2.0_PlatformIO\.pio\libdeps\my_board\Adafruit BusIO)
|   |-- Wire @ 1.0 (License: Unknown, Path: C:\Users\micha\.platformio\packages\framework-arduino-avr\libraries\Wire)
|   |-- SPI @ 1.0 (License: Unknown, Path: C:\Users\micha\.platformio\packages\framework-arduino-avr\libraries\SPI)
|-- Wire @ 1.0 (License: Unknown, Path: C:\Users\micha\.platformio\packages\framework-arduino-avr\libraries\Wire)
|-- Adafruit GFX Library @ 1.11.5 (License: Unknown, Path: C:\Users\micha\OneDrive\Berglerwerk\Berglerwerk_Github\Pen2.0\Pen2.0_PlatformIO\.pio\libdeps\my_board\Adafruit GFX Library)
|   |-- Adafruit BusIO @ 1.14.1 (License: Unknown, Path: C:\Users\micha\OneDrive\Berglerwerk\Berglerwerk_Github\Pen2.0\Pen2.0_PlatformIO\.pio\libdeps\my_board\Adafruit BusIO)
|   |   |-- Wire @ 1.0 (License: Unknown, Path: C:\Users\micha\.platformio\packages\framework-arduino-avr\libraries\Wire)
|   |   |-- SPI @ 1.0 (License: Unknown, Path: C:\Users\micha\.platformio\packages\framework-arduino-avr\libraries\SPI)
|   |-- Wire @ 1.0 (License: Unknown, Path: C:\Users\micha\.platformio\packages\framework-arduino-avr\libraries\Wire)
|   |-- SPI @ 1.0 (License: Unknown, Path: C:\Users\micha\.platformio\packages\framework-arduino-avr\libraries\SPI)
|-- TFT 22 ILI9225 @ 1.4.5 (License: Unknown, Path: C:\Users\micha\OneDrive\Berglerwerk\Berglerwerk_Github\Pen2.0\Pen2.0_PlatformIO\.pio\libdeps\my_board\TFT 22 ILI9225)
|   |-- SPI @ 1.0 (License: Unknown, Path: C:\Users\micha\.platformio\packages\framework-arduino-avr\libraries\SPI)
|-- ArduinoJson @ 6.21.2 (License: Unknown, Path: C:\Users\micha\OneDrive\Berglerwerk\Berglerwerk_Github\Pen2.0\Pen2.0_PlatformIO\.pio\libdeps\my_board\ArduinoJson)
Building in release mode
MethodWrapper(["checkprogsize"], [".pio\build\my_board\firmware.elf"])
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [=         ]  12.4% (used 318 bytes from 2560 bytes)
Flash: [===       ]  34.2% (used 9808 bytes from 28672 bytes)
.pio\build\my_board\firmware.elf  :

section                     size      addr

.data                        110   8388864

.text                       9698         0

.bss                         208   8388974

.comment                      17         0

.note.gnu.avr.deviceinfo      64         0

.debug_aranges               224         0

.debug_info                 3857         0

.debug_abbrev               2550         0

.debug_line                 1180         0

.debug_str                   881         0

Total                      18789
<lambda>(["upload"], [".pio\build\my_board\firmware.hex"])
AVAILABLE: custom
CURRENT: upload_protocol = custom
avrdude -CC:\Users\micha\.platformio\packages/tool-avrdude/avrdude.conf -patmega32u4 -cflip1 -s -U flash:w:.pio\build\my_board\firmware.hex:i 

avrdude: Warning: USB bDeviceClass = 255 (expected 254)
avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.00s

avrdude: Device signature = 0x1e9587 (probably m32u4)
avrdude: Error: "lfuse" memory not accessible using FLIP
avrdude: Error: "lfuse" memory not accessible using FLIP
avrdude: NOTE: "flash" memory has been specified, an erase cycle will be performed
         To disable this feature, specify the -D option.
avrdude: erasing chip
avrdude: reading input file ".pio\build\my_board\firmware.hex"
avrdude: writing flash (9808 bytes):

Writing | ################################################## | 100% 0.56s

avrdude: 9808 bytes of flash written
avrdude: verifying flash memory against .pio\build\my_board\firmware.hex:
avrdude: load data flash data from input file .pio\build\my_board\firmware.hex:
avrdude: input file .pio\build\my_board\firmware.hex contains 9808 bytes
avrdude: reading on-chip flash data:

Reading | ################################################## | 100% 0.17s

avrdude: verifying ...
avrdude: 9808 bytes of flash verified

avrdude: Error: "lfuse" memory not accessible using FLIP
avrdude: Error: "lfuse" memory not accessible using FLIP
avrdude: Error: "lfuse" memory not accessible using FLIP
avrdude: Error: "lfuse" memory not accessible using FLIP
avrdude: safemode: Sorry, reading back fuses was unreliable. I have given up and exited programming mode

avrdude done.  Thank you.

*** [upload] Error 1

**Reply by MCUdude** [reply]
This is from the Avrdude 6.3 manual:

```
-u 
Disables the default behaviour of reading out the fuses three times before pro- gramming, then verifying at the end of programming that the fuses have not changed. If you want to change fuses you will need to specify this option, as avrdude will see the fuses have changed (even though you wanted to) and will change them back for your "safety". This option was designed to prevent cases of fuse bits magically changing (usually called safemode).
If one of the configuration files contains a line
default_safemode = no;
safemode is disabled by default. The ‘-u’ option’s effect is negated in that case,
i. e. it enables safemode.
Safemode is always disabled for AVR32, Xmega and TPI devices.

-s 
Disable safemode prompting. When safemode discovers that one or more fuse bits have unintentionally changed, it will prompt for confirmation regarding whether or not it should attempt to recover the fuse bit(s). Specifying this flag disables the prompt and assumes that the fuse bit(s) should be recovered without asking for confirmation first.
```

So you should probably try `-u` instead of `-s` in your case

**Comment by michbichler** [comment]
Perfect, there are no error anymore! Thank you so much for your help!

**Comment by michbichler** [comment]
To make my custom board appear as a COM port in a serial terminal program, I installed the USB Serial (CDC) driver. However, I can no longer flash the board. Do I have to change anything in platformio.ini?

![Screenshot](https://github.com/avrdudes/avrdude/assets/114282671/6b09617c-0a34-4554-a7a1-23bad5fc0138)
